resourceIndexer/coreStagingS3.js

```diff
@@ -0,0 +1,292 @@
+import {logger} from './logger.js';
+import { SOURCE, VALID_AUSPICE_PATTERNS, INVALID_AUSPICE_PATTERNS,
+  DATESTAMP_REGEX, MAIN_DATASET_JSON } from './constants.js';
+
+/**
+ * The inventory of buckets (especially the core bucket) is in some ways a
+ * historical record of work over the years, but this isn't really what we want
+ * to display to users. As some examples:
+ *    - Files which don't match a resource to list should be excluded
+ *    - Datestampted files (i.e. _YYYY-MM-DD in the filename) are excluded
+ *      (we use S3 versioning instead)
+ * 
+ * If the s3 object is to be excluded we return false here.
+ * 
+ * In the case where the object represents a (part of) a resource we want to
+ * expose, then we categorise it here by adding the following properties:
+ *   - source (STAGING or CORE)
+ *   - resourceType (dataset, narrative or intermediate)
+ *   - id (the ID by which objects will be grouped together.
+ *        For datasets this is the nextstrain.org URL path, without any temporal signifier)
+ *   - subresourceType (currently only for resourceType=dataset)
+ */
+function categoriseCoreObjects(item, staging) {
+  const key = item.key
+  item.source = staging ? SOURCE.STAGING : SOURCE.CORE;
+  item.baseUrl = `https://nextstrain-${staging?'staging':'data'}.s3.amazonaws.com/${key}`
+  if (key.startsWith('search_')
+    || key.startsWith('manifest_')
+    || key.startsWith('datasets_')
+  ) return false;
+
+  // On the core bucket, directory-like hierarchies are used for intermediate
+  // files only (not auspice files). There are other resourceTypes here we may
+  // consider in the future -- e.g. model output JSONs
+  if (key.includes("/")) {
+    if (staging===true) return false;
+    if (key.startsWith('files/')) {
+      if (
+        key.includes('/archive/')
+        || key.includes('/test/')
+        || key.includes('/branch/')
+        || key.includes('/trial/')
+        || key.includes('/test-data/')
+        || key.includes('jen_test/')
+        || key.match(/\/\d{4}-\d{2}-\d{2}_results.json/) // forecasts-ncov
+        || key.endsWith('.png')                          // forecasts-ncov
+      ) {
+        return false;
+      }
```

Comment:
My rationale here is that broadly applicable and reused constants (e.g. those used in auspice file) go in `constants.js`, but these patters are only applicable to the core bucket, and only used in one place. I feel like this is a more suitable location for these ad-hoc patterns.

Reply:
Looking at the data again today, it may be enough to invert this and only take intermediate files in the `./files/` "directory". I'll revisit this tomorrow.